float8_experimental/fsdp_utils.py

```diff
@@ -0,0 +1,61 @@
+import math
+import warnings
+from typing import List
+
+import torch
+import torch.nn as nn
+from float8_experimental.float8_dynamic_utils import WeightWithDynamicFloat8CastTensor
+from float8_experimental.float8_linear import Float8Linear
+from float8_experimental.float8_linear_utils import linear_requires_sync
+from float8_experimental.float8_utils import EPS
+
+
+def precompute_float8_scale_for_fsdp(module: nn.Module) -> None:
+    """
+    Calculate scale for all float8 parameters after optimizer step
+    It performs a single all-reduce instead of many all-reduces for each parameter
+    Exmaple usage:
+        model(input).sum().backward()
+        optim.step()
+        precompute_float8_scale_for_fsdp(model)
+    """
+    from torch.distributed._tensor import DTensor
+
+    if any(
+        isinstance(m, Float8Linear)
+        and linear_requires_sync(
+            m.scaling_type_x, m.scaling_type_w, m.scaling_type_dL_dY
+        )
+        for m in module.modules()
+    ):
+        raise NotImplementedError("Only supports delayed scaling")
+    float8_linears: List[Float8Linear] = [
+        m
+        for m in module.modules()
+        if isinstance(m, Float8Linear)
+        and isinstance(m.weight, DTensor)
+        and isinstance(m.weight._local_tensor, WeightWithDynamicFloat8CastTensor)
+    ]
+    weights: List[DTensor] = [float8_linear.weight for float8_linear in float8_linears]
+
+    def compute_scales(weights: List[DTensor]):
```

Comment:
optional nit: maybe move outside to prevent nested functions?

Reply:
curious what is the downside of nested functions

@weifengpy By the way, this was originally a nested function just so that we could try to `torch.compile` it effectively in the `scales = compute_scales(weights)` line. Does it still need to be a separate function for `torch.compile` reasons? If so, we should probably add a comment before the `def compute_scales` mentioning that it is separate for `torch.compile`; otherwise, we can consider inlining the function.

Reply:
I will remove nested functions to make the code easy to read. I profiled the unit test and `precompute_float8_scale_for_fsdp` takes 1.9ms. thta's a tiny portion of the overall training loop. No obvious reason to speed up with `torch.compile` yet. I can bring back the nested function in case we need `torch.compile` again.

<img width="491" alt="Screenshot 2024-07-11 at 1 50 46 PM" src="https://github.com/pytorch-labs/float8_experimental/assets/134637289/aaf29ef0-56d1-4004-a7da-c6af6ea3b54b">